transports/tls/src/verifier.rs

```diff
@@ -0,0 +1,217 @@
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! TLS 1.3 certificates and handshakes handling for libp2p
+//!
+//! This module handles a verification of a client/server certificate chain
+//! and signatures allegedly by the given certificates.
+
+use crate::certificate;
+use rustls::{
+    client::{HandshakeSignatureValid, ServerCertVerified, ServerCertVerifier},
+    internal::msgs::handshake::DigitallySignedStruct,
+    server::{ClientCertVerified, ClientCertVerifier},
+    Certificate, DistinguishedNames, SignatureScheme,
+};
+
+/// Implementation of the `rustls` certificate verification traits for libp2p.
+///
+/// Only TLS 1.3 is supported. TLS 1.2 should be disabled in the configuration of `rustls`.
+pub struct Libp2pCertificateVerifier;
+
+/// libp2p requires the following of X.509 server certificate chains:
+///
+/// - Exactly one certificate must be presented.
+/// - The certificate must be self-signed.
+/// - The certificate must have a valid libp2p extension that includes a
+///   signature of its public key.
+impl Libp2pCertificateVerifier {
+    /// Return the list of SignatureSchemes that this verifier will handle,
+    /// in `verify_tls12_signature` and `verify_tls13_signature` calls.
+    ///
+    /// This should be in priority order, with the most preferred first.
+    pub fn verification_schemes() -> Vec<SignatureScheme> {
+        vec![
+            // TODO SignatureScheme::ECDSA_NISTP521_SHA512 is not supported by `ring` yet
+            SignatureScheme::ECDSA_NISTP384_SHA384,
+            SignatureScheme::ECDSA_NISTP256_SHA256,
+            // TODO SignatureScheme::ED448 is not supported by `ring` yet
+            SignatureScheme::ED25519,
+            // In particular, RSA SHOULD NOT be used unless
+            // no elliptic curve algorithms are supported.
+            SignatureScheme::RSA_PSS_SHA512,
+            SignatureScheme::RSA_PSS_SHA384,
+            SignatureScheme::RSA_PSS_SHA256,
+            SignatureScheme::RSA_PKCS1_SHA512,
+            SignatureScheme::RSA_PKCS1_SHA384,
+            SignatureScheme::RSA_PKCS1_SHA256,
+        ]
+    }
+}
+
+impl ServerCertVerifier for Libp2pCertificateVerifier {
+    fn verify_server_cert(
```

Comment:
So we can dial with a PK and without. If we dial with a PK then we should get a cert.extension.public_key and check whether they are equal. This code was mistakenly removed in 3b1ff348999da1543.

cc @marten-seemann

Reply:
From specs:

```
The public host key allows the peer to calculate the peer ID of the peer it is connecting to. Clients MUST verify that the peer ID derived from the certificate matches the peer ID they intended to connect to, and MUST abort the connection if there is a mismatch.
```

Reply:
PK = public key.

> The libp2p extension needs to be present in the certificate in any case.

Yes but if we dial with a known PK we MUST verify that the PK in the extension equals to the PK we are dialing to. Is it correct?

Reply:
That's correct. Although I'd expect the more common case to be that you know the Peer ID in advance, not the public key (for peer IDs where the key is not inlined).

Reply:
Verification that we connected to the intended peer (`/p2p`) happens in the `Swarm`, independent of the given `Transport` in `rust-libp2p`. I think we are covering this requirement for any `Transport` as long as people are using the `Swarm` abstraction[^1].


[^1]: Most users are as that is what a lot of functionality is built on top of but in theory, you could use it without `libp2p-swarm`. I know of one project that does and there, the check is implemented manually.